site/_posts/2018-02-14-the-pathname-to-discovering-a-memory-leak-in-ruby.html.md

```diff
@@ -0,0 +1,429 @@
+that, the jobs that caused the leak seemed to be at random, and instrumenting
+the code to determine the leak was hard to do in a generic fashion. That is not
+to say we didn't try wrap our workers with some common Ruby profiling tools (we
+did... many times...), but the result usually led us to wonder if were actually
+dealing with the ["Observer effect"][7] and introducing new problems instead of
+"observing" existing ones.
+
+Although it took much more time for it to be truly observed (around a month's
+time), we also found that the `MIQ Server` (our worker monitor process) was
+leaking as well.  Since it had a more predictable workflow, this was the better
+candidate to do some deeper process of elimination testing.  It took some time
+to get a scenario where this was leaking consistently, but once we did it was
+simpler to do A/B testing with it.
+
+
+Step 1.5:  Reproducing the issue... fasterer
+--------------------------------------------
+
+One problem with using the `MiqServer` process as the target for observing the
+leak is it took _**days**_/_**weeks**_ to determine if it was leaking or just
+the process taking it's sweet time time to stabilize it's memory usage.  This
+led to very long gaps between making a change and having observable results to
+view from it.
+
+One of the least invasive things that we eventually did was decrease the server
+`:monitor_poll` and `:worker_monitor_frequency` values from `"5.seconds"` and
+`"15.seconds"` respectively to `"1.second"` for both.  This decreased the wait
+time between each iteration of it's normal work and thankfully increased the
+leak rate.
+
+We also increased the frequency of normal method calls by doing things like
+`40.times { do_some_work }` in places where we thought the leak might be, and
+if there was rapid growth, we knew where to probe deeper.
+
+The suggestions in this step happened for us throughout the process of
+discovering the cause of the leak, and didn't necessarily happen in between
+this step and the next (hence the "Step 1.5").  Steps to accelerate the
+feedback loop, such as this, should be done as early into the memory leak
+debugging process as possible.
+
+
+Step 2:  Narrowing down the culprits
+------------------------------------
+
+As mentioned previously, the `MiqServer` process has a pretty regular work
+loop.  In case you aren't familiar with the workflow of the `MiqServer` work
+loop, here is a gist of the important steps:
+
+1. heartbeat (tell other servers that I am still alive)
+2. process any messages sent to the server process over the message queue
+3. check on the health status of other server processes (and log them)
+4. check the status of all of the workers processes (stop/start them as needed)
+5. fetch queue work to assign to worker processes
+
+Each of these steps happen in varying intervals and happen on a mostly
+consistent basis, baring any database latency fluctuations or worker changes.
```

Comment:
Typo `baring` -> `barring`.  Unless you meant `baring` in that it exposed those things?

Reply:
You are correct, thanks!